### PR TITLE
Fixes #2884 by conditionally compiling in the surround_with/container_doc call dependent on the Elixir version.

### LINKED ISSUE #2884: Warning Inspect.Algebra.surround_many/5 is deprecated on Elixir 1.8.0-otp-21.2.2
### Environment

* Elixir version (elixir -v):
    ```elixir
    Erlang/OTP 21 [erts-10.2.1] [source] [64-bit] [smp:2:2] [ds:2:2:10] [async-threads:1] [hipe]
    
    Elixir 1.8.0 (compiled with Erlang/OTP 21)
    ```
* Database and version (PostgreSQL 9.4, MongoDB 3.2, etc.): N/A
* Ecto version (mix deps): `ecto 3.0.6 (Hex package) (mix)`
* Database adapter and version (mix deps):  N/A
* Operating system: RedHat Linux

### Current behavior

Fresh compile with the dependency of `{:ecto, "~> 3.0.6"},` results in:
```elixir
==> ecto
Compiling 54 files (.ex)
warning: Inspect.Algebra.surround_many/5 is deprecated. Use Inspect.Algebra.container_doc/6 instead
Found at 3 locations:
  lib/ecto/changeset.ex:2755
  lib/ecto/query/inspect.ex:16
  lib/ecto/query/inspect.ex:30

Generated ecto app
```

### Expected behavior

Clean compile.

### PR DESCRIPTION
**Pull request by OvermindDL1**
The files are not `mix format`'d as they were not formatted to start with, however the unquote branch and call itself is formatted according to `mix format`.

The tests have been run and still pass:
```elixir
╰─➤  mix test
Compiling 16 files (.ex)
.........................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................                                                                                                       

Finished in 15.6 seconds
75 doctests, 894 tests, 0 failures

Randomized with seed 364475
```

In addition manually checking the inspection results also appear identical.  The default separator is the one already used by the original `surround_with/5` call thus there should be no functionality changes.

Any other changes needed?  :-)

### CODE REVIEW

**Comment by OvermindDL1** [comment]
I also tested on a few different 1.8.0 versions including the release, release candidates, and a beta (master at the time).

File: lib/ecto/query/inspect.ex [hunk 1]
```diff
@@ -27,7 +33,13 @@ defimpl Inspect, for: Ecto.Query do
...
     end)
 
-    surround_many("#Ecto.Query<", list, ">", opts, fn str, _ -> str end)
+    # TODO: Replace unquote with just `container_doc` when Elixir < 1.8.0 is no longer supported
+    unquote(
+      if(Version.match?(System.version(), "~> 1.8.0-dev"),
+        do: quote(do: :container_doc),
+        else: quote(do: :surround_many)
+      )
+    )("#Ecto.Query<", list, ">", opts, fn str, _ -> str end)
```
Comment (josevalim) [inline lib/ecto/query/inspect.ex]: Can you please move the logic to outside of the function? Something like this:

```
    container_doc =  if Version.match?(System.version(), ">= 1.6.0"), do: :container_doc, else: :surround_many
```

Note the proper version to check is >= 1.6.0 and you don't need to quote atoms. Thanks!

Comment (OvermindDL1) [reply lib/ecto/query/inspect.ex]: Sure, also as a note, doing `">= 1.6.0"` caused it to fail on the RC.1 and so forth, are those essentially pretended to not exist then?  If so that's fine, just want to be explicit.  :-)

Comment (josevalim) [reply lib/ecto/query/inspect.ex]: > Sure, also as a note, doing ">= 1.8.0" caused it to fail on the RC.1 and so forth

RC.1 of what? This function exists since v1.6.0 :)

Comment (OvermindDL1) [reply lib/ecto/query/inspect.ex]: Pushed those changes with the `">= 1.8.0"` ignoring test 1.8.0 versions.  :-)

Comment (josevalim) [reply lib/ecto/query/inspect.ex]: Sorry, I think something is getting lost in translation. Let me try again: why not use `>= 1.6.0` (one-six-oh) instead of `>= 1.8.0` (one-eight-oh)?

Comment (OvermindDL1) [reply lib/ecto/query/inspect.ex]: Oh was it introduced in 1.6.0?!  Ah so it was!  I was running a *really* old version of elixir on the CI until recently...  ^.^;

Updating!  Few minutes...  :-)

File: lib/ecto/query/inspect.ex [hunk 2]
```diff
@@ -13,7 +15,8 @@ defimpl Inspect, for: Ecto.Query.DynamicExpr do
     inspected =
       Inspect.Ecto.Query.expr(expr, List.to_tuple(names), %{expr: expr, params: params})
 
-    surround_many("dynamic(", [Macro.to_string(binding), inspected], ")", opts, fn str, _ -> str end)
+    # TODO: Replace unquote with just `container_doc` when Elixir < 1.8.0 is no longer supported
```
Comment (josevalim) [inline lib/ecto/query/inspect.ex]:
```suggestion
    # TODO: Replace unquote with just `container_doc` when Elixir < 1.6.0 is no longer supported
```

File: lib/ecto/changeset.ex [hunk 3]
```diff
@@ -2747,12 +2747,15 @@ end
...
 
+  container_doc = if(Version.match?(System.version(), ">= 1.6.0"), do: :container_doc, else: :surround_many)
+
   def inspect(changeset, opts) do
     list = for attr <- [:action, :changes, :errors, :data, :valid?] do
       {attr, Map.get(changeset, attr)}
     end
 
-    surround_many("#Ecto.Changeset<", list, ">", opts, fn
+    # TODO: Replace unquote with just `container_doc` when Elixir < 1.8.0 is no longer supported
```
Comment (josevalim) [inline lib/ecto/changeset.ex]:
```suggestion
    # TODO: Replace unquote with just `container_doc` when Elixir < 1.6.0 is no longer supported
```

File: lib/ecto/changeset.ex [hunk 4]
```diff
@@ -2747,12 +2747,15 @@ end
 defimpl Inspect, for: Ecto.Changeset do
   import Inspect.Algebra
 
+  container_doc = if(Version.match?(System.version(), ">= 1.6.0"), do: :container_doc, else: :surround_many)
```
Comment (josevalim) [inline lib/ecto/changeset.ex]:
```suggestion
  container_doc = if Version.match?(System.version(), ">= 1.6.0"), do: :container_doc, else: :surround_many
```

File: lib/ecto/query/inspect.ex [hunk 5]
```diff
@@ -3,6 +3,8 @@ import Kernel, except: [to_string: 1]
 
 alias Ecto.Query.{DynamicExpr, JoinExpr, QueryExpr}
 
+container_doc = if(Version.match?(System.version(), ">= 1.6.0"), do: :container_doc, else: :surround_many)
```
Comment (josevalim) [inline lib/ecto/query/inspect.ex]:
```suggestion
container_doc = if Version.match?(System.version(), ">= 1.6.0"), do: :container_doc, else: :surround_many
```

File: lib/ecto/query/inspect.ex [hunk 6]
```diff
@@ -27,7 +30,8 @@ defimpl Inspect, for: Ecto.Query do
         string
     end)
 
-    surround_many("#Ecto.Query<", list, ">", opts, fn str, _ -> str end)
+    # TODO: Replace unquote with just `container_doc` when Elixir < 1.8.0 is no longer supported
```
Comment (josevalim) [inline lib/ecto/query/inspect.ex]:
```suggestion
    # TODO: Replace unquote with just `container_doc` when Elixir < 1.6.0 is no longer supported
```

**Comment by OvermindDL1** [comment]
Ah sorry, I'm used to putting parenthesis when using an `if` like a single-line ternery for explicitness.  ^.^;

Updated it and comments.  :-)

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
